### PR TITLE
ENH: Use flake8 to check for PEP8 violations in doctests

### PR DESCRIPTION
**Pull request by FHaase**
- [x] tests added 
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Improvement to PR #23381 regarding issue #23154 

Enables `--doctests` flag, resulting in doctests being checked by flake8
Uses [flake8.api.legacy](http://flake8.pycqa.org/en/latest/user/python-api.html) to invoke the tests.

Downsides are:
* > Flake8 3.0.0 presently does not have a public, stable Python API.
* provided options in `flake8.get_style_guide` don't have an effect. Thus `setup.cfg` had to be enhanced.
* returns the count of each error within the complete file, not just the requested function. [Might be confusing]

`python ./scripts/validate_docstrings.py pandas.read_excel`
```
                [...]
                Parameter "convert_float" description should finish with "."
        Errors in doctest sections
                2 F721 syntax error in doctest
                2 F821 undefined name '_make_signature'
```

Help in the form of `Use "flake8 pandas/util/_decorators.py" for further information.` could be added to the output:
```
pandas/util/_decorators.py:320:26: F721 syntax error in doctest
pandas/util/_decorators.py:321:13: F721 syntax error in doctest
pandas/util/_decorators.py:322:15: F821 undefined name '_make_signature'
pandas/util/_decorators.py:322:31: F821 undefined name 'f'
```

### CODE REVIEW

**Comment by gfyoung** [comment]
@FHaase : Thanks for the PR!  Given your first comment in the description, I would recommend opening an issue first.  That's generally what we would recommend instead of jumping right to the PR.

**Comment by datapythonista** [comment]
Thanks @FHaase, this looks much better. Can you mention the issue you close, and the other PR related to this in the description please.

I just saw couple of things. Doesn't feel like we should be using a module named legacy. I guess you checked it, but doesn't flake8 have a newer module if they call that one legacy?

Also, feels like when we validate a docstring, we validate all the pep8 problems in the file, right? We should find a way to validate a single docstring.

When this is done, we should also implement tests for this, as the rest of validations.

In any case, good work, this will be very helpful once is done.

**Comment by FHaase** [comment]
> I just saw couple of things. Doesn't feel like we should be using a module named legacy. I guess you checked it, but doesn't flake8 have a newer module if they call that one legacy?

[`flake8.api`](https://gitlab.com/pycqa/flake8/tree/master/src/flake8/api) is empty besides the legacy submodule

> Also, feels like when we validate a docstring, we validate all the pep8 problems in the file, right? We should find a way to validate a single docstring.

Yes I mentioned it. The only way of fixing this might be to create a temp-file just with the docstring in it and run flake8 on that?

File: setup.cfg [hunk 1]
```diff
@@ -28,6 +28,31 @@ exclude =
...
+doctests = True
+#TODO fix doctests
+exclude_from_doctest =
+    ./pandas/_libs
+    ./pandas/api
+    ./pandas/compat
+    ./pandas/computation
+    ./pandas/core
+    ./pandas/errors
+    ./pandas/io
```
Comment (jreback) [inline setup.cfg]: a number of these have been removed recently, can you edit

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by FHaase** [comment]
> I just saw couple of things. Doesn't feel like we should be using a module named legacy. I guess you checked it, but doesn't flake8 have a newer module if they call that one legacy?

Dug down into the code and basically recreated the legacy api.

File: scripts/tests/test_validate_docstrings.py [hunk 2]
```diff
@@ -164,6 +165,8 @@ def contains(self, pat, case=True, na=np.nan):
 
         Examples
         --------
+        >>> import pandas as pd
+        >>> import numpy as np
```
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: As numpy is a dependency of pandas, it makes more sense to me to import numpy first.

File: scripts/tests/test_validate_docstrings.py [hunk 3]
```diff
@@ -584,6 +587,29 @@ def prefix_pandas(self):
         pass
 
 
+class BadExamples(object):
+
+    def doctest(self):
```
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: I think `doctest` is not a very descriptive name if this test is to check whether pep8 should fail because of unknown `np`. The rest of the example is very verbose to just check that. I'd use the description to explain what is this test about. And in the examples just with `>>> np.nan` could be enough.

Then, couple more tests could be added, to test the typical pep8 issues we have, like `>>> foo = 1+2` (missing spaces)...

File: scripts/validate_docstrings.py [hunk 4]
```diff
@@ -24,6 +24,11 @@
 import inspect
 import importlib
 import doctest
+import tempfile
+from contextlib import contextmanager
+
+from flake8.main.application import Application as Flake8
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: I think it complicates things unnecessarily importing objects other than modules. `import contextlib` and `import flake8.main.application` seems like the simplest and best way. And then use accordinly `@contextlib.contextmanager` and `app = flake8.main.application.Application`. This way, just by reading those lines it's obvious what is the module and what is being used, and there is not need to go to the imports and start decoding aliases.

File: scripts/validate_docstrings.py [hunk 5]
```diff
@@ -331,6 +336,41 @@ def parameter_mismatches(self):
...
+            application.run_checks([file.name])
+            application.report()
+            stats = application.guide.stats
+            return [
+                "{} {} {}".format(s.count, s.error_code, s.message)
+                for s in stats.statistics_for('')
+            ]
+
+    @contextmanager
+    def _file_representation(self):
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: This name is not very descriptive. I think this is the examples as temporary file, so something like `_examples_as_temp_file` seems much clearer.

Not sure if it's worth having two separate methods, I'd generate the file in the method that validates pep8. But if we have two methods, I'd have this first, as this is being used by the other (`pep8_violations`).

File: scripts/validate_docstrings.py [hunk 6]
```diff
@@ -331,6 +336,41 @@ def parameter_mismatches(self):
...
+        """
+        Temporarily creates file with current function inside.
+        The signature and body are **not** included.
+
+        :returns file
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: we use numpydoc, not this syntax

File: scripts/validate_docstrings.py [hunk 7]
```diff
@@ -331,6 +336,41 @@ def parameter_mismatches(self):
...
+        """
+        create_function = 'def {name}():\n' \
+                          '    """{doc}"""\n' \
+                          '    pass\n'
+
+        name = self.name.split('.')[-1]
+        lines = self.clean_doc.split("\n")
+        indented_lines = [(' ' * 4) + line if line else ''
+                          for line in lines[1:]]
+        doc = '\n'.join([lines[0], *indented_lines])
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: I'm a bit confused with this. I guess I'm missing something. If in our docstring we have something like:
```
Some desc.

Examples
--------
>>> val = 1+2
>>> print(val)
```

I'd expect to have in the temporary file to validate:

```
val = 1+2
print(val)
```

As we won't validate the description, or anything else. I think numpydoc returns the lines with code, so writing to the file should be a single line of code. Is there any advantage generating a function?

File: scripts/validate_docstrings.py [hunk 8]
```diff
@@ -446,7 +486,7 @@ def validate_one(func_name):
         if doc.summary != doc.summary.lstrip():
             errs.append('Summary contains heading whitespaces.')
         elif (doc.is_function_or_method
-                and doc.summary.split(' ')[0][-1] == 's'):
+              and doc.summary.split(' ')[0][-1] == 's'):
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: it'd be better to not do unrelated changes

File: scripts/validate_docstrings.py [hunk 9]
```diff
@@ -331,6 +336,41 @@ def parameter_mismatches(self):
...
+        with self._file_representation() as file:
+            application = Flake8()
+            application.initialize(["--doctests", "--quiet"])
+            application.run_checks([file.name])
+            application.report()
+            stats = application.guide.stats
+            return [
+                "{} {} {}".format(s.count, s.error_code, s.message)
+                for s in stats.statistics_for('')
+            ]
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: I'd personally prefer to `yield from application.guide.stats.statistics_for('')` instead of returning the string. In case the information is used in more than one place, having the components can be useful.

Also, it's an opinion, but I wouldn't use a property here, as this feels more like an action than an attribute. A method `validate_pep8` seems easier to understand.

**Review by datapythonista — COMMENTED** [review]
Added some comments, mainly about code clarity and simplicity. But looks good, I think we're almost there.

File: scripts/validate_docstrings.py (hunk 7 above)
Comment (FHaase) [reply scripts/validate_docstrings.py]: Well flake8 supports doctests by adding `--doctests` as an argument. And the Examples follow the doctest standard.

> Examples
> An optional section for examples, using the doctest format. 

Having to carve out the examples is not trivial and would require writing multiple tests. So creating a file with a dummy-function is a simple hack, getting ugly cause it has to restore the indentation of the docstring. One test of the `BadExamples` class ensures it is working.

File: scripts/tests/test_validate_docstrings.py (hunk 3 above)
Comment (FHaase) [reply scripts/tests/test_validate_docstrings.py]: The reason to use flake8 to check for pep8 issues was to not have to write all these tests. flake8 should be tested enough to work properly. 

This test basically only verifies that flake8 is actually testing the docstrings.

File: scripts/validate_docstrings.py [hunk 10]
```diff
@@ -331,6 +335,26 @@ def parameter_mismatches(self):
...
+        lines = (' ' * 4 + line if line else '' for line in self.examples)
+        examples = '\n'.join(lines)
+        with tempfile.NamedTemporaryFile(mode='w') as file:
+            func = create_function.format(name=name, examples=examples)
+            file.write(func)
+            file.flush()
+
+            application = flake8.Application()
+            application.initialize(["--doctests", "--quiet"])
+            application.run_checks([file.name])
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: can you just leave the writing to the file and this line inside the context manager, and creating the app object, initializing... have it outside?

File: scripts/validate_docstrings.py [hunk 11]
```diff
@@ -490,6 +514,13 @@ def validate_one(func_name):
         for param_err in param_errs:
             errs.append('\t{}'.format(param_err))
 
+    pep8_errs = [error for error in doc.validate_pep8()]
+    if pep8_errs:
+        errs.append('Errors in doctests')
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: Not exactly "errors" I'd say:

```suggestion
        errs.append('Linting issues in doctests:')
```

File: scripts/validate_docstrings.py [hunk 12]
```diff
@@ -490,6 +514,13 @@ def validate_one(func_name):
         for param_err in param_errs:
             errs.append('\t{}'.format(param_err))
 
+    pep8_errs = [error for error in doc.validate_pep8()]
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: I think this is easier to read:
```suggestion
    pep8_errs = list(doc.validate_pep8())
```

**Review by datapythonista — COMMENTED** [review]
Looking great. Still few comments. And we'll probably have to wait to #23161 to reuse the code extraction from the examples, but it should be merged very soon.

File: scripts/tests/test_validate_docstrings.py [hunk 13]
```diff
@@ -703,10 +748,23 @@ def test_bad_generic_functions(self, func):
...
          ('pandas.Series.rename in `See Also` section '
-          'does not need `pandas` prefix',))
+          'does not need `pandas` prefix',)),
+        # Examples tests
+        ('BadExamples', 'numpy_imported',
+         ('F811 It\'s assumed that pandas and numpy'
+          ' are imported as pd or np',)),
+        ('BadExamples', 'pandas_imported',
+         ('F811 It\'s assumed that pandas and numpy'
+          ' are imported as pd or np',)),
```
Comment (datapythonista) [inline scripts/tests/test_validate_docstrings.py]: these two examples need to be removed

File: scripts/validate_docstrings.py [hunk 14]
```diff
@@ -331,6 +336,33 @@ def parameter_mismatches(self):
...
         return errs
 
+    @contextlib.contextmanager
+    def _write_examples_code_to_temp_file(self):
+        """
+        Generate file with source code from examples section.
+        """
+        content = ''.join(('import numpy as np; '
+                           'import pandas as pd  # noqa: F401,E702\n',
+                           *self.examples_source_code))
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: wouldn't it be simpler to join by `\n` instead, and remove the semicolon and the noqa?

File: scripts/validate_docstrings.py [hunk 15]
```diff
@@ -331,6 +336,33 @@ def parameter_mismatches(self):
...
+                application = flake8.main.application.Application()
+                application.initialize(["--quiet"])
+                application.run_checks([file.name])
+                application.report()
+
+            for statistic in application.guide.stats.statistics_for(''):
+                if statistic.message.endswith('from line 1'):
+                    statistic.message = "It's assumed that pandas and numpy" \
+                                        " are imported as pd or np"
+                yield statistic
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: I guess with the original `yield from` we may get a pep8 error of duplicate import, or something similar. I'm fine with that. Besides the pep8 errors, we'll have the error of the other section implemented in #23161, which should make it very easy for users to identify the problem.

So, I'd simply leave the `yield from ...`

File: scripts/validate_docstrings.py [hunk 16]
```diff
@@ -331,6 +336,33 @@ def parameter_mismatches(self):
...
+            file.flush()
+            yield file
+
+    def validate_pep8(self):
+        if self.examples:
+            with self._write_examples_code_to_temp_file() as file:
+                application = flake8.main.application.Application()
+                application.initialize(["--quiet"])
+                application.run_checks([file.name])
+                application.report()
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: Sorry I wasn't clear in my previous review about this. This is what I'd do:
```
application = flake8.main.application.Application()
application.initialize(["--quiet"])
with tempfile.NamedTemporaryFile(mode='w') as f:
    f.write(content)
    f.flush()
    application.run_checks([f.name])
application.report()
```

I think it's a good practice to have inside the context manager, the code that requires to be there.

Also, as this file is going quite significantly, I think it's better to keep all this functionality in a single function, so each validation is somehow isolated. I think in another context it makes more sense what you did of splitting, but I don't think we'll reuse the temporary file creation, and I think the improvement in simplicity makes it worth.

File: scripts/validate_docstrings.py [hunk 17]
```diff
@@ -331,6 +336,33 @@ def parameter_mismatches(self):
...
+        content = ''.join(('import numpy as np; '
+                           'import pandas as pd  # noqa: F401,E702\n',
+                           *self.examples_source_code))
+        with tempfile.NamedTemporaryFile(mode='w') as file:
+            file.write(content)
+            file.flush()
+            yield file
+
+    def validate_pep8(self):
+        if self.examples:
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: small thing, but my preference is having:
```
if not self.examples:
    return
```
this avoids having the rest of the function indented, and in my opinion increases readability.

**Review by datapythonista — COMMENTED** [review]
Looks quite good, added a couple of comments about what to expect from #23161, and some style preferences, but we're almost there.

I'm really looking forward to have this merged. After that I'll add this validation to the CI, so we don't need to keep reviewing the pep8 of examples manually, which is a significant amount of work.

Thanks for all the work on this.

File: scripts/validate_docstrings.py (hunk 14 above)
Comment (FHaase) [reply scripts/validate_docstrings.py]: * regarding E702: with one line the great thing was to achieve intercepting of the error message.
* regarding F401: All examples not using both: np and pd would fail because of an unused import.

The implementation of `self.examples_source_code` of PR #23161 returns an array with `\n` at the end, so joining with `\n` isn't possible.

File: scripts/validate_docstrings.py (hunk 15 above)
Comment (FHaase) [reply scripts/validate_docstrings.py]: I found `F811 redefinition of unused 'np' from line 1` could be confusing when read first.

File: scripts/validate_docstrings.py [hunk 18]
```diff
@@ -407,6 +411,26 @@ def examples_source_code(self):
         lines = doctest.DocTestParser().get_examples(self.raw_doc)
         return [line.source for line in lines]
 
+    def validate_pep8(self):
+        if not self.examples:
+            return
+
+        content = ''.join(('import numpy as np  # noqa: F401\n',
```
Comment (datapythonista) [inline scripts/validate_docstrings.py]: may be we can add a comment before this line explaining why the `F401` is required.

**Review by datapythonista — APPROVED** [review]
looks great, thanks @FHaase 

Just one suggestion, but ready to be merged on green I think. @jreback you requested changed, if you want to take a look.

**Review by jreback — APPROVED** [review]
merge when satisfied

**Comment by datapythonista** [comment]
Thanks @FHaase, extremely useful PR

File: scripts/validate_docstrings.py (hunk 18 above)
Comment (FHaase) [reply scripts/validate_docstrings.py]: Well the content basically never shows as it's only existing within tmp files. And if someone thinks it's unnecessary the tests would fail. 
So I don't think commenting is needed there.

Comment (datapythonista) [reply scripts/validate_docstrings.py]: The comment was not for the temp file, but our code, and I think it's useful to know what the `F401` is about when reading this part of code (without having to remove it and see that fails). Anyway, this is merged now. :)
